pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -628,6 +643,103 @@ func maintainKubernetesVersion(log logr.Logger, kubernetesVersion string, autoUp
+	var (
+		maintenanceResults                    = make(map[string]updateResult)
+		sshKeypairRotationEnabled             = v1beta1helper.IsSSHKeypairAutoRotationEnabled(shoot)
+		observabilityPasswordsRotationEnabled = v1beta1helper.IsObservabilityAutoRotationEnabled(shoot)
+		etcdEncryptionKeyRotationEnabled      = v1beta1helper.IsETCDEncryptionKeyAutoRotationEnabled(shoot)
+		etcdEncryptionKeyRotationPhase        = v1beta1helper.GetShootETCDEncryptionKeyRotationPhase(shoot.Status.Credentials)
+	)
+
+	if sshKeypairRotationEnabled && v1beta1helper.ShootEnablesSSHAccess(shoot) &&
+		sshKeypairRotationPassedRotationPeriod(shoot, now.Time, *shoot.Spec.Maintenance.AutoRotation.Credentials.SSHKeypair.RotationPeriod) {
+		reason := "Automatic rotation of SSH keypair configured"
+		log.Info("SSH keypair for workers will be rotated", "reason", reason)
+		maintenanceResults[v1beta1constants.ShootOperationRotateSSHKeypair] = updateResult{
+			description:  "SSH keypair rotation started",
+			reason:       reason,
+			isSuccessful: true,
+		}
+	}
+
+	if observabilityPasswordsRotationEnabled &&
+		observabilityPasswordsRotationPassedRotationPeriod(shoot, now.Time, *shoot.Spec.Maintenance.AutoRotation.Credentials.Observability.RotationPeriod) {
+		reason := "Automatic rotation of observability passwords configured"
+		log.Info("Observability passwords will be rotated", "reason", reason)
+		maintenanceResults[v1beta1constants.OperationRotateObservabilityCredentials] = updateResult{
+			description:  "Observability passwords rotation started",
+			reason:       reason,
+			isSuccessful: true,
+		}
+	}
+
+	if etcdEncryptionKeyRotationEnabled &&
+		etcdEncryptionKeyRotationPassedRotationPeriod(shoot, now.Time, *shoot.Spec.Maintenance.AutoRotation.Credentials.ETCDEncryptionKey.RotationPeriod) {
+		if len(etcdEncryptionKeyRotationPhase) == 0 || etcdEncryptionKeyRotationPhase == gardencorev1beta1.RotationCompleted {
+			reason := "Automatic rotation of etcd encryption key configured"
+			log.Info("ETCD Encryption key will be rotated", "reason", reason)
+			maintenanceResults[v1beta1constants.OperationRotateETCDEncryptionKey] = updateResult{
+				description:  "ETCD Encryption key rotation started",
+				reason:       reason,
+				isSuccessful: true,
+			}
+		} else {
+			reason := "ETCD encryption key rotation is already in progress"
+			maintenanceResults[v1beta1constants.OperationRotateETCDEncryptionKey] = updateResult{
+				description:  "Could not start ETCD encryption key rotation",
+				reason:       reason,
+				isSuccessful: false,
+			}
+		}
+	}
+
+	return maintenanceResults
+}
+
+// sshKeypairRotationPassedRotationPeriod checks if the rotation period for ssh keypair has passed.
+func sshKeypairRotationPassedRotationPeriod(shoot *gardencorev1beta1.Shoot, now time.Time, period metav1.Duration) bool {
+	credentialsCreationTime := shoot.CreationTimestamp.Time
```

Comment:
The shoot creation timestamp isn't exactly the credentials creation time. This is just an easy-to-do estimate. 
Can you add a comment about this and that it is necessary for shoots that still did not have their credentials rotated once.
Also perhaps rename the variable to `latestRotationCompletionTime`:

```suggestion
    // If the shoot has just been created or the credentials have never been rotated, use the shoot's creation timestamp to determine whether the rotation period has passed.
	latestRotationCompletionTime := shoot.CreationTimestamp.Time
```

```